pkg/virt-handler/ratelimitcmd/executor_test.go

```diff
@@ -0,0 +1,82 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2022 Red Hat, Inc.
+ *
+ */
+
+package ratelimitcmd_test
+
+import (
+	"fmt"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"k8s.io/apimachinery/pkg/util/clock"
+
+	"kubevirt.io/kubevirt/pkg/virt-handler/ratelimitcmd"
+)
+
+var testsExecError = fmt.Errorf("error occurred")
+
+var _ = Describe("rate-limited executor with limited-exponential-backoff rate-limiter", func() {
+
+	const clockStepOffset = 1 * time.Second
+
+	var (
+		testsClock *clock.FakeClock
+		executor   *ratelimitcmd.RateLimitedExecutor
+	)
+
+	BeforeEach(func() {
+		testsClock = clock.NewFakeClock(time.Now())
+		executor = ratelimitcmd.NewRateLimitedExecutor(
+			ratelimitcmd.NewFixedExponentialLimitedBackoffWithClock(ratelimitcmd.DefaultMaxStep, testsClock))
+
+		testsClock.Step(time.Nanosecond)
+	})
+
+	It("should execute command immediately on first execution", func() {
+		Expect(executor.Exec(failingCommandStub())).To(MatchError(testsExecError))
+	})
+
+	It("should not execute command before the rate-limiter step end time", func() {
+		Expect(executor.Exec(failingCommandStub())).To(MatchError(testsExecError))
+		Expect(executor.Exec(failingCommandStub())).To(Succeed())
+	})
+
+	It("should execute command after the rate-limiter step end time passed", func() {
+		Expect(executor.Exec(failingCommandStub())).To(MatchError(testsExecError))
+
+		Expect(executor.Exec(failingCommandStub())).To(Succeed())
+		testsClock.SetTime(executor.StepEnd().Add(clockStepOffset))
```

Comment:
I like less the fact that you need to expose an inner value of an wrapped attriubute.. two layers deep the implementation stack. It introduces a harsh dependency on the implementation.

It will be much better to expose the values and factors of the backoff, in order to calculate here the needed growth of the clock vs doing this.
```
firstStepDuration := 3.1
secondStepDuration := firstStepDuration*1.8,
```
Something like that...

Reply:
Done